src/Bicep.Core/TypeSystem/DeployTimeConstantContainerVisitor.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class DeployTimeConstantContainerVisitor : SyntaxVisitor
+    {
+        private readonly SemanticModel semanticModel;
+
+        private readonly List<SyntaxBase> deployTimeConstantContainers = new();
+
+        public DeployTimeConstantContainerVisitor(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        public static IEnumerable<SyntaxBase> CollectDeployTimeConstantContainers(SemanticModel semanticModel)
+        {
+            var visitor = new DeployTimeConstantContainerVisitor(semanticModel);
+
+            visitor.Visit(semanticModel.Root.Syntax);
+
+            return visitor.deployTimeConstantContainers;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            if (syntax.TryGetKeyText() is { } propertyName &&
+                this.semanticModel.Binder.GetParent(syntax) is { } objectSyntax &&
+                this.semanticModel.TypeManager.GetDeclaredType(objectSyntax) is ObjectType objectType &&
```

Comment:
Feels like it would be useful to extract this functionality (getting a property type from a property) into an extension method.

Reply:
Good pt. Extracted this into `ObjectPropertySyntax.TryGetTypeProperty`.